### ISSUE TITLE
if we use GRPO and args.kl_coef is non-zero, is the KL computation incorrect?

### ISSUE THREAD

**Issue by adol001** [THUDM/slime#397]
slime/backends/megatron_utils/loss.py

```
    if args.kl_coef == 0:
        # when kl_coef is 0, we won't compute ref_log_prob
        xs = log_probs if log_probs is not None else values
        kl = [torch.zeros_like(x, dtype=torch.float32, device=x.device) for x in xs]
    else:
        kl = [
            compute_approx_kl(
                log_probs[i],
                ref_log_probs[i],
                kl_loss_type=args.kl_loss_type,
            )
            for i in range(len(log_probs))
        ]

    if args.advantage_estimator in ["grpo", "gspo"]:
        rewards = torch.tensor(rewards, dtype=torch.float32, device=kl[0].device)
        returns = get_grpo_returns(rewards, kl)
        # TODO: is the copy necessary?
        advantages = [r for r in returns]

```

slime/utils/ppo_utils.py
```
def get_grpo_returns(
    rewards: torch.Tensor,
    kl: list[torch.Tensor],
):
    returns = []
    for i in range(len(rewards)):
        returns.append(torch.ones_like(kl[i]) * rewards[i])
    return returns
```

If we use GRPO and args.kl_coef is non-zero, then kl_coef is effectively not used; the reward isn’t adjusted by args.kl_coef

**Comment by primorLee** [comment]
I checked this again against current `main` (624b824a). The mismatch is still present:

- `--kl-coef` is documented as the KL coefficient for reward shaping.
- A nonzero value loads the reference model and computes per-token KL in `compute_advantages_and_returns`.
- The GRPO/GSPO/CISPO branch then calls `get_grpo_returns(rewards, kl)`, but that helper only broadcasts the scalar reward and never consumes either `kl` or `args.kl_coef`.

So the reference forward is paid for, but changing `--kl-coef` does not change GRPO returns or advantages.

There are two materially different fixes, and I do not want to choose an objective on behalf of the project:

1. apply per-token reward shaping, e.g. `reward - beta * KL_t`, in the GRPO-family return/advantage path; or
2. reject `--kl-coef != 0` for GRPO-family estimators and require `--use-kl-loss --kl-loss-coef` for a KL term.

Which behavior matches slime's intended objective? I can prepare a focused CPU-tested PR once that contract is confirmed.

**Comment by YeonwooSung** [comment]
Opened https://github.com/THUDM/slime/pull/2359 to reject `--kl-coef != 0` for GRPO-style estimators (`grpo` / `gspo` / `cispo`) and point users at `--use-kl-loss` / `--kl-loss-coef`.

This follows the maintainer decision on #399 that GRPO should use `kl_loss` rather than applying KL to rewards. `get_grpo_returns()` is unchanged.
